pyccel/parser/semantic.py

```diff
@@ -1526,6 +1534,9 @@ def _visit_PyccelAdd(self, expr, **settings):
         atoms = [a['is_pointer'] or a['is_target'] for a in atoms if a['rank']>0]
         args  = [self._visit(a, **settings) for a in expr.args]
 
+        if None in args:
+            args = [a if a is not None else Integer(0) for a in args]
+
```

Comment:
in which case we get `None` ?

Reply:
In the case where the code is wrong and variables have not been declared. See tests/errors/semantic for examples

Reply:
I think this would also be the case if the variable was declared with an expression that pyccel does not understand

Reply:
then we shouldn't modify the code to ignore it ,it will crash at printing time maybe

Reply:
It crashes at various places in the code when it tries to interact with it. It is one of the reasons why our error messages are so unintelligible. In order to improve our errors it is important that it exits gracefully rather than crashing out.

We discussed previously that errors during each stage shouldn't stop that stage completing in order to find as many of the errors associated with that stage as possible. If the code crashes when we try to check something such as the dtype then this is no longer true.

If a `None` is returned then an error is also added to Errors(). If there are errors then the execution stops. If the execution will be stopped then the result won't be generated so it doesn't matter if the output would have been wrong.

We are therefore asking the rest of Semantic to ignore it but as we have an associated Error we are not ignoring it completely

Reply:
there are some blocking errors that we have no choice but to stop the process and one of them is the type inference if we can't because of None, then we should stop, modifying the code is more dangerous because we may not detect it at all

Reply:
Ok. So `get_variable` (line 310 of parser/semantic.py) should never return None.
It should populate Errors with a Fatal error and raise an error if it finds nothing?

Reply:
we only continue if it's not a blocking error and we report it  so there is no risk because we are reporting it, we also need to always think about the performance, adding conditions will make the parsing slower only if it's necessary then we have no choice

Reply:
`get_variable` is different because we can cases like this one  https://github.com/pyccel/pyccel/blob/master/pyccel/parser/semantic.py#L1396.
if we don't have that kind of cases then sure we can raise an error there

Reply:
Currently none of the errors in pyccel are blocking due to this line (parser/base.py:257):
```
        # flag for blocking errors. if True, an error with this flag will cause
        # Pyccel to stop
        # TODO ERROR must be passed to the Parser __init__ as argument

        self._blocking = False
```

It seems strange to me that blocking or not depends on the parser rather than the error.

I would have expected fatal errors to be blocking and normal errors to be non-blocking

Reply:
yes maybe then we should make them blocking